### PR TITLE
Fix QPY deserialization of UCRX, UCRY, and UCRZ gates

### LINKED ISSUE #7847: QPY can't serialize `UCRZGate`
### Environment

- **Qiskit Terra version**: 0.19.2
- **Python version**: 3.9.7
- **Operating system**: macOS Monterey 12.2.1

### What is happening?

While developing a [Grover's tutorial](https://github.com/Qiskit/qiskit-ibm-runtime/pull/236) for the new Qiskit Runtime primitives, I found a bug where the grover circuit submitted to runtime provider with sampler program return `TypeError: __init__() takes 2 positional arguments but 5 were given`. After some investigation, I have narrowed down the issue to `UCRZGate` which is part of the [Grover operator](https://qiskit.org/documentation/stubs/qiskit.circuit.library.GroverOperator.html#qiskit.circuit.library.GroverOperator) in the grover circuit.

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [1], in <cell line: 20>()
     18 circuit_json = json.loads(json_str)
     19 qpy_file = io.BytesIO(base64.b64decode(circuit_json["circuits"]))
---> 20 circuit = qpy_serialization.load(qpy_file)[0]

File /usr/local/anaconda3/envs/primitives/lib/python3.9/site-packages/qiskit/circuit/qpy_serialization.py:1776, in load(file_obj)
   1774 circuits = []
   1775 for _ in range(file_header[5]):
-> 1776     circuits.append(_read_circuit(file_obj, qpy_version))
   1777 return circuits

File /usr/local/anaconda3/envs/primitives/lib/python3.9/site-packages/qiskit/circuit/qpy_serialization.py:1912, in _read_circuit(file_obj, version)
   1910 custom_instructions = _read_custom_instructions(file_obj, version, vectors)
   1911 for _instruction in range(num_instructions):
-> 1912     _read_instruction(file_obj, circ, out_registers, custom_instructions, version, vectors)
   1913 for vec_name, (vector, initialized_params) in vectors.items():
   1914     if len(initialized_params) != len(vector):

File /usr/local/anaconda3/envs/primitives/lib/python3.9/site-packages/qiskit/circuit/qpy_serialization.py:1121, in _read_instruction(file_obj, circuit, registers, custom_instructions, version, vectors)
   1119         elif gate_name in {"BreakLoopOp", "ContinueLoopOp"}:
   1120             params = [len(qargs), len(cargs)]
-> 1121         gate = gate_class(*params)
   1122     gate.condition = condition_tuple
   1123 if label_size > 0:

TypeError: __init__() takes 2 positional arguments but 5 were given
```

### How can we reproduce the issue?

```python
import io
import json
import base64
from numpy import pi

from qiskit.circuit import qpy_serialization
from qiskit import QuantumCircuit

qc = QuantumCircuit(3)
qc.ucrz([0,0,0,-pi], [1,2], [0])
qc.measure_all()

buf = io.BytesIO()
qpy_serialization.dump(qc, buf)
json_str = json.dumps({
    'circuits': base64.b64encode(buf.getvalue()).decode('utf8')
})
circuit_json = json.loads(json_str)
qpy_file = io.BytesIO(base64.b64decode(circuit_json["circuits"]))
circuit = qpy_serialization.load(qpy_file)[0]
```
qpy code based on [this answer](https://quantumcomputing.stackexchange.com/questions/21795/is-there-a-quantum-circuit-builder-package-that-allows-you-to-import-export-a-ci) by @mtreinish on StackExchange.
### What should happen?

Circuits containing `UCRZGate` should be able to be serialized by QPY.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes the deserialization of the UCRX, UCRY, and UCRZ gates.
These gates aren't constructed in the standard way as other gate objects
where the parameters are passed directly as args to the class
constructor. Instead they behave the same as Initialize and the params
are passed as a list to the constructor to make a new gate object. This
commit fixes the handling in deserialization to treat these gates the
same as Initialize so that it can be properly deserialized by
qpy.load().

### Details and comments

Fixes #7847

### CODE REVIEW

File: releasenotes/notes/ucr-gates-qpy-b8f6fb1e34fae258.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  - |
+    Fixed an issue with QPY deserialization via the :func:`.qpy.load` function
+    of the :class:`~.UCRXGate`, :class:`~.UCRYGate`, and :class:`~.UCRZGate`.
+    Previously, a QPY file that contained any of these gates would error
+    when trying to load the file.
+    Fixed `#7847 <https://github.com/Qiskit/qiskit-terra/issues/7847>`__
+issues:
+  - |
+    QPY deserialization with the :func:`.qpy.load` function of a directly
+    instantiated :class:`~.UCPauliRotGate` object in a circuit will fail
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply releasenotes/notes/ucr-gates-qpy-b8f6fb1e34fae258.yaml:12]: The only place I've seen it directly used is by `UCRX`, ``UCRY`, and `UCRZ` as the parent class. It's basically just a base class for the axis specific versions to de-duplicate the decomposition of the gate. As to how to fix it the only way I can think to is to bump the qpy version to v5 and add a field to the custom instruction type for `UCPauliRotGate` to store the rotation angle as an extra field in the payload. There is also the hacky equivalent where we modify `_write_instruction` to add the rotation axis to the parameter list for a `UCPauliRotGate` and handle that in a custom path on deserialization but that feels like the wrong way to handle this. Ideally we'd refactor all these classes to store the angle list as a single parameter and then for `UCPauliRotGate` we have a second parameter for the rotation angle, but that's a larger refactor.

**Review by nkanazawa1989 — APPROVED** [review]
LGTM

**Review by HuangJunye — COMMENTED** [review]
In order to completely fixed this issue for sending the circuit to IBM Cloud / IBM Quantum, we also need to update the QPY version on the respective provider as well, don't we? https://github.com/Qiskit/qiskit-ibm-runtime/pull/250#issuecomment-1084917371

**Comment by mtreinish** [comment]
> In order to completely fixed this issue for sending the circuit to IBM Cloud / IBM Quantum, we also need to update the QPY version on the respective provider as well, don't we? [Qiskit/qiskit-ibm-runtime#250 (comment)](https://github.com/Qiskit/qiskit-ibm-runtime/pull/250#issuecomment-1084917371)

Generally, yes to fix an issue in submitting the circuit to the qiskit runtime service you'll need to backport this fix to the qiskit-ibm-runtime package's fork of qpy. However, in this case I don't think it is necessary because deserialization is done server side and the error would occur in the qiskit-runtime service's environment itself and the not the client when it tried to load the qpy file. When the runtime environment updates their qiskit version to 0.20.1 with this fix I would expect it to just start working because there is no change in the file format as part of this fix.

**Comment by mtreinish** [comment]
Sigh, it looks like @nkanazawa1989's approval isn't enough despite being a codeowner for qpy because the tests for qpy don't live in a path that matches qpy/ in the codeowners file. We should move the tests to fix this in the future.

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
@Mergifyio requeue

**Comment by jakelishman** [comment]
@Mergifyio refresh

**Comment by jakelishman** [comment]
Oh duh, it wasn't re-evaluating because it didn't have the automerge label.

**Pull request merged**